File: sem4/Języki skryptowe - scripting languages/list5/java/Graph.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

public class Graph
{
    private class Edge
    {
        Node from;
        Node to;
        int weight;

        Edge(Node from, Node to, int dist)
        {
            this.from = from;
            this.to = to;
            this.weight = dist;
        }

        @Override
        public String toString()
        {
            return ((from.toString()) + " --> " + (to.toString()) + " [" + weight + "]");
        }
        public void setWeight(int i)
        {
            this.weight = i;
        }
    }
    private class Node implements Comparable<Node>
    {
        String name;

        Node(String name)
        {
            this.name = name;
        }

        @Override
        public String toString()
        {
            return this.name;
        }
        @Override
        public int compareTo(Node n)
        {
            return this.name.compareTo(n.name);
        }
    }

    private ArrayList<Edge> edges;
    private ArrayList<Node> nodes;

    public Graph()
    {
        this.edges = new ArrayList<>();
        this.nodes = new ArrayList<>();
    }

    public int addNode(String name)
    {
        boolean unique = true;
        for(Node node : nodes)
            if(node.toString().equals(name))
            {
                unique = false;
                break;
            }
        if(unique)
        {
            Node tmp = new Node(name);
            nodes.add(tmp);
            return nodes.indexOf(tmp);
        }
        else return -1;
    }
    public int addEdge(Node from, Node to, int weight)
    {
        if (from == null || to == null || from.equals(to))
            return -1;
        for(Edge e : edges)
            if (e.from.compareTo(from) == 0 && e.to.compareTo(to) == 0)
                return -1;
        Edge tmp = new Edge(from, to, weight);
        edges.add(tmp);
        return edges.indexOf(tmp);
    }
    public boolean removeEdge(int idx)
    {
       if(edges.remove(idx) == null)
           return false;
       else
           return true;
    }
    public boolean removeNode(String name)
    {
        for(Node node : nodes)
            if(node.toString().equals(name))
            {
                nodes.remove(node);
                edges.removeIf(edge -> (edge.from.equals(node) || edge.to.equals(node)));
                return true;
            }
        return false;
    }

    public boolean changeNodeName(String oldName, String newName)
    {
        Node temp = getNode(oldName);
        if(temp != null)
        {
            temp.name = newName;
            return true;
        }
        return false;
    }
    public boolean changeEdgeWeight(int idx, int weight)
    {
        try
        {
            getEdge(idx).weight = weight;
            return true;
        }
        catch(Exception e)
        {
            return false;
        }
    }


    public Node getNode(String name)
    {
        for(int i=0; i<nodes.size(); i++)
            if(nodes.get(i).toString().equals(name))
                return nodes.get(i);
        return null;
    }
    public Edge getEdge(int idx)
    {
        return edges.get(idx);
    }

    public void showNodes()
    {
        System.out.println("----- nodes -----");
        for(Node node : nodes)
            System.out.println(node.toString());
    }
    public void showEdges()
    {
        System.out.println("----- edges -----");
        System.out.println("idx. from --> to [weight]");
        for(int i = 0; i < edges.size(); i++)
            System.out.println(i + ".\t" + edges.get(i).toString());
        System.out.println("-----------------");
    }
    public void getGraphInfo()
    {
        showNodes();
        showEdges();
    }

    public void DFS(Node start_node)
    {
        System.out.println("Nodes visited in DFS");
        Stack stack = new Stack();
        boolean[] visited = new boolean[nodes.size()];

        stack.push(start_node);
        for(boolean b : visited)
            b = false;

        while(!stack.isEmpty())
            DFS_visit((Node)stack.pop(), stack, visited);

    }
    private void DFS_visit(Node node, Stack stack, boolean[] visited)
    {
        if(!visited[nodes.indexOf(node)])
        {
            System.out.println(node.toString());
            //all the stuff we wanna do with the visited node happens here
        }
        visited[nodes.indexOf(node)] = true;
        for(Edge e : edges)
            if(e.from.equals(node) && !visited[nodes.indexOf(e.to)])
                stack.push(e.to);
    }

    public void BFS(Node start_node)
    {
        System.out.println("Nodes visited in BFS");
        Queue<Node> queue = new LinkedList<Node>();
        boolean[] visited = new boolean[nodes.size()];

        queue.add(start_node);
        for(boolean b : visited)
            b = false;

        while(!queue.isEmpty())
            BFS_visit(queue.poll(), queue, visited);

    }
    private void BFS_visit(Node node, Queue queue, boolean[] visited)
    {
        if(!visited[nodes.indexOf(node)])
        {
            System.out.println(node.toString());
            //all the stuff we wanna do with the visited node happens here
        }
        visited[nodes.indexOf(node)] = true;
        for(Edge e : edges)
            if(e.from.equals(node) && !visited[nodes.indexOf(e.to)] && !queue.contains(e.to))
                queue.add(e.to);
    }
}

File: sem4/Języki skryptowe - scripting languages/list5/python/graph.py
import subprocess

from py4j.java_gateway import JavaGateway


class Graph:
    def __init__(self):
        subprocess.Popen(['java', 'GraphEntry'])
        self.gateway = JavaGateway()
        self.graphs = {}
        self.currentGraph = None

    def add_graph(self, name):
        self.graphs[name] = self.gateway.entry_point.createGraph()

    def choose_graph(self, name):
        if self.graphs.__contains__(name):
            self.currentGraph = self.graphs[name]
            return True
        return False

    def show_graphs(self):
        print(self.graphs)

    def add_edge(self, node_from, node_to, distance):
        return self.currentGraph.addEdge(self.get_node(node_from), self.get_node(node_to), distance)

    def remove_edge(self, idx):
        return self.currentGraph.removeEdge(idx)

    def change_edge_weight(self, idx, weight):
        return self.currentGraph.changeEdgeWeight(idx, weight)

    def add_node(self, name):
        return self.currentGraph.addNode(name)

    def remove_node(self, name):
        return self.currentGraph.removeNode(name)

    def get_node(self, name):
        return self.currentGraph.getNode(name)

    def change_node_name(self, old, new):
        return self.currentGraph.changeNodeName(old, new)

    def show_current_graph_nodes(self):
        self.currentGraph.showNodes()

    def show_current_graph_edges(self):
        self.currentGraph.showEdges()

    def show_current_graph_info(self):
        self.currentGraph.getGraphInfo()

    def print_current_graph_nodes_DFS(self, start_node):
        self.currentGraph.DFS(start_node)

    def print_current_graph_nodes_BFS(self, start_node):
        self.currentGraph.BFS(start_node)

    def stop(self):
        self.gateway.close()
        self.gateway.shutdown()
